Guide/file-storage.markdown

```diff
@@ -575,3 +575,244 @@ config = do
+            else do
+                -- Image style not found, so create it.
+                let options :: StoreFileOptions = def
+                        { directory = imageStylePathDirectory
+                        , preprocess = applyImageMagick "jpg" ["-resize", cs size <> "^", "-gravity", "center", "-extent", cs size, "-quality", "85%", "-strip"]
+                        }
+
+                storedFile <- storeFileFromPath (cs $ storagePrefix <> originalImageDirectory <> "/" <> uuid) options
+
+                renderFile (cs $ storagePrefix <> storedFile.path) "application/jpg"
+
+-- | Extracts the directory and UUID from a path like "pictures/8ed22caa-11ea-4c45-a05e-91a51e72558d"
+extractDirectoryAndUUID :: (?context :: context, ConfigProvider context) => Text -> (Text, Text)
+extractDirectoryAndUUID inputText =
+    case reverse parts of
+        uuid : pathSegments -> (T.intercalate "/" (reverse pathSegments), uuid)
+        _ -> ("", "")
+    where
+        frameworkConfig = ?context.frameworkConfig
+        trimmedText = T.replace (frameworkConfig.baseUrl <> "/") "" inputText
+        parts = T.splitOn "/" trimmedText
+```
+
+Now, from any `Show` action, we can use the image style. Here we create a 400px x 200px image style for the original image.
+
+```haskell
+[hsx|<img src={pathTo $ RenderImageStyleAction 400 200 imageUrl} />|]
+where
+    imageUrl = "http://localhost:8000/static/picture/b4c8f55c-16d6-41f0-9503-77352b134e14"
+```
+
+### Security Considerations
+
+The above implementation is not secure. Anyone can request any image style of any image. This may put a burden on the site. Imagine a malicious user requesting thousands of different dimensions of a 100MB image. The server would have to generate thousands of images. This would put a lot of load on the server. To prevent this, we need to add some access control. The way we would do it, as that is the simplest way, is to add a signed token to the image style URL. The token would be generated by the server and would contain the original image path, the desired width and height. The server would then verify the token and only serve the image style if the token is valid. This way, the server would not have to generate the image style on the fly, but only if the token is valid. If the token is not valid, then the server would return an access denied.
+
+We start by adding the `cryptonite` package to the `default.nix` file, and re-start `./start` to get the package.
+
+```nix
+# default.nix
+haskellEnv = import "${ihp}/NixSupport/default.nix" {
+    ihp = ihp;
+    haskellDeps = p: with p; [
+        cabal-install
+        base
+        # ...
+        cryptonite # <--- ADD THIS
+    ];
+```
+
+Then we need to generate a private and public key pair. We can do it in our Config file, and save the keys for later use.
+
+```haskell
+-- Config/Config.hs
+import Crypto.PubKey.RSA as RSA
+
+newtype RsaPublicAndPrivateKeys = RsaPublicAndPrivateKeys (RSA.PublicKey, RSA.PrivateKey)
```

Comment:
I think this could also be a data structure:

```suggestion
data RsaPublicAndPrivateKeys = RsaPublicAndPrivateKeys { publicKey :: RSA.PublicKey, privatekey :: RSA.PrivateKey }
```

Reply:
Should I move this and the below suggestions to `Application.Helper.Controller`. Or maybe  `Application.Helper.RSA` ?

Reply:
Kept in Application.Helper.Controller

Reply:
I'd keep it in Config as it's only used for configuration purposes